lib/src/clojure_lsp/feature/java_interop.clj

```diff
@@ -120,16 +120,24 @@
         (update :analysis merge analysis)
         (db/upsert-cache! db))))
 
-(defn download-and-analyze! [{:keys [db]}]
-  (let [jdk-download-url (settings/get db [:java :jdk-source-download-uri] default-jdk-source-download-uri)
-        global-cache-dir (config/global-lsp-cache-dir)
+(defn ^:private download-jdk-source
+  [jdk-source-uri db]
+  (let [global-cache-dir (config/global-lsp-cache-dir)
         dest-jdk-file (io/file global-cache-dir "jdk")
         dest-jdk-result-file (io/file dest-jdk-file "successfully-downloaded")]
     (if (shared/file-exists? dest-jdk-result-file)
       (logger/info "[JDK] Skipping downloading JDK source, already found at" dest-jdk-file)
       (do
-        (download-jdk! jdk-download-url dest-jdk-file)
-        (spit dest-jdk-result-file "1")))
-    (logger/info "[JDK] Analyzing JDK source...")
-    (analyze-jdk-source! (.getCanonicalPath dest-jdk-file) db)
-    (logger/info "[JDK] Source analyzed successfully.")))
+       (download-jdk! jdk-source-uri dest-jdk-file)
+       (spit dest-jdk-result-file "1")))
+    (.getCanonicalPath dest-jdk-file) db))
+
+(defn retrieve-and-analyze! [{:keys [db]}]
+  (let [jdk-source-uri (settings/get db [:java :jdk-source-uri] default-jdk-source-uri)
+        download-jdk-source? (settings/get db [:java :download-jdk-source?] false)]
+    (when-let [jdk-source-uri' (or (and (shared/plain-uri? jdk-source-uri)
+                                        (shared/filename->uri jdk-source-uri db))
+                                   (when download-jdk-source? (download-jdk-source jdk-source-uri db)))]
+      (logger/info (str "[JDK] Analyzing JDK source found at " jdk-source-uri "..."))
+      (analyze-jdk-source! jdk-source-uri' db)
+      (logger/info (str "[JDK] Source found at " jdk-source-uri " analyzed successfully.")))))
```

Comment:
We already mention `jdk-source-uri` on line 141, I think we don't need to mention again?

Reply:
np, will remove.